rustup-init.ps1

```diff
@@ -0,0 +1,174 @@
+        --default-toolchain <default-toolchain>    Choose a default toolchain to install
+'@
+}
+
+function main([AllowEmptyCollection()] [string[]]$Argv = @()) {
+    foreach ($arg in $Argv) {
+        if ("$arg" -eq "-h" -or "$arg" -eq "--help") {
+            Write-Usage
+            return
+        }
+    }
+
+    Install-Rustup $Argv
+}
+
+function Get-Architecture {
+    $platform = [System.Environment]::OSVersion.Platform
+    switch -Wildcard ($platform) {
+        # Linux and Darwin both report as Unix
+        "Unix" {
+            $ostype = uname -s
+            $cputype = uname -m
+            break
+        }
+        "Win*" {
+            $ostype = (Get-WmiObject Win32_OperatingSystem).Caption
+            $cputype = (Get-WmiObject Win32_OperatingSystem).OSArchitecture
+            break
+        }
+        default {
+            Write-Error "could not determine OS platform: $platform"
+        }
+    }
+
+    switch -Wildcard ($ostype) {
+        "Darwin" {
+            $ostype = "apple-darwin"
+            break
+        }
+        "Linux" {
+            $ostype = "unknown-linux-gnu"
+            break
+        }
+        "*Windows*" {
+            $ostype = "pc-windows-msvc"
+            break
+        }
+        default {
+            Write-Error "unrecognized OS type: $ostype"
+        }
+    }
+
+    switch ($cputype) {
+        # Reported by Windows 32-bit operating systems
+        "32-bit" {
+            $cputype = "i686"
+            break
+        }
+        # Reported by Windows 64-bit operating systems
+        "64-bit" {
+            $cputype = "x86_64"
+            break
+        }
+        # Reported by Darin/Linux 64-bit operating systems
+        "x86_64" {
+            break
+        }
+        default {
+            Write-Error "unknown CPU type: $cputype"
+        }
+    }
+
+    $arch = "${cputype}-${ostype}"
+    Write-Debug "Detected arch '$arch'"
+    return "$arch"
+}
+
+# If RUSTUP_UPDATE_ROOT is unset, default it.
+function Get-RustupUpdateRoot {
+    if (Test-Path env:\RUSTUP_UPDATE_ROOT) {
+        return "$env:RUSTUP_UPDATE_ROOT"
+    } else {
+        return "https://static.rust-lang.org/rustup"
+    }
+}
+
+function Install-Rustup([AllowEmptyCollection()] [string[]]$Argv) {
+    $arch = Get-Architecture
+    switch -Wildcard ($arch) {
+        "*windows*" { $ext = ".exe"; break }
+        default { $ext = "" }
+    }
+
+    $url = "$(Get-RustupUpdateRoot)/dist/$arch/rustup-init$ext"
+
+    $dir = New-TemporaryDirectory
+    $file = Join-Path "$dir" "rustup-init$ext"
+
+    try {
+        if ("$arch" -match "windows") { Set-HighestEncryption }
+
+        Write-Debug "Downloading $url to $file"
+        (New-Object System.Net.WebClient).DownloadFile($url, $file)
+        if (-Not ("$arch" -match "windows")) { chmod u+x "$file" }
+
+        Write-Debug "Calling '$file $Argv'"
+        & "$file" @Argv
+    } finally {
+        Write-Debug "Cleaning $dir"
```

Comment:
nit: this is verbose not debug

Reply:
Ah, even better. That would be a `Write-Verbose` I presume?

Reply:
Correct.